### PR TITLE
Add Projects API 

### PR DESCRIPTION
**Pull request by Nitish145**
#### Describe the changes you have made in this PR -

1. Adds the following projects endpoints:
    -  `api/v1/projects` to fetch all public or user accessible projects.
    -  `api/v1/users/:id/projects` to fetch user specific projects.
    -  `api/v1/projects/:id` to GET, UPDATE & DELETE project details.
    -  `api/v1/projects/featured` to fetch featured projects.
    -  `api/v1/projects/:id/toggle-star` to star/unstar a project.
    -  `api/v1/projects/:id/fork` to fork a project.

2. Add project/projects schema.

3. Add specs for each of the projects controller actions.

### The updated `docs` are available [here](https://nitish145.github.io/slate/)

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by Nitish145** [comment]
@tachyons please check!!

File: app/controllers/api/v1/projects_controller.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  before_action :view_access?, only: [:show]
+  before_action :sanitize_name, only: [:update]
+  before_action :set_options, except: [:destroy, :toggle_star]
+
+  SORTABLE_FIELDS = ["view", "created_at"]
+
+  # GET /api/v1/projects
+  def index
+    if request.headers["Authorization"].present?
+      authenticate_user!
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:19]: Move this logic to base controller and set @current_user there

File: app/controllers/api/v1/projects_controller.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      @projects = Project.where(
+        "author_id = ? OR  project_access_type = ?", @current_user.id, "Public")
+    else
+      @projects = Project.where("project_access_type = ?", "Public")
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:23]:
```suggestion
      @projects = Project.public
```

File: app/controllers/api/v1/projects_controller.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      Arel.sql(SortingHelper.sort_fields(params[:sort], SORTABLE_FIELDS))
+    ) if params.has_key?(:sort)
+    @options[:links] = link_attrs(paginate(@projects), api_v1_projects_url)
+
+    render json: Api::V1::ProjectSerializer.new(paginate(@projects), @options)
+  end
+
+  # GET /api/v1/users/:id/projects/
+  def user_projects
+    if @current_user.eql?(User.find(params[:id]))
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:37]:
```suggestion
    if @current_user.id == params[:id]
```

File: app/controllers/api/v1/projects_controller.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      @projects = Project.where(author_id: params[:id])
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:38]: Should be @user.projects

File: app/controllers/api/v1/projects_controller.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    end
+  end
+
+  # /api/v1/projects/:id/fork
+  def create_fork
+    if @current_user.id == @project.author_id
+      api_error(status: 409, errors: "Cannot fork your own project!")
+    else
+      @project_new = @project.dup
+      @project_new.update!(
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:92]: Can move this to `Project#fork!`

File: .rubocop.yml [hunk 6]
```diff
@@ -32,6 +32,11 @@ Metrics/MethodLength:
   Exclude:
     - 'spec/**/*' # not a big deal in spec helper methods
 
+Metrics/ClassLength:
+  Max: 100
+  Exclude:
+    - 'app/models/**/*' #fat models and slim controllers principle
```
Comment (tachyons) [inline .rubocop.yml:38]: We don't follow fat models, we want both controllers and model to be slim as possible

File: app/controllers/api/v1/base_controller.rb [hunk 7]
```diff
@@ -45,7 +45,10 @@ def authenticate_user
   end
 
   def authenticate_user!
-    authenticate_user || UnauthenticatedError
+    raise MissingAuthHeader if request.headers["Authorization"].blank?
+
+    authenticate_user
+    raise UnauthenticatedError if @current_user.nil?
```
Comment (tachyons) [inline app/controllers/api/v1/base_controller.rb:51]:
```suggestion
    raise UnauthenticatedError if current_user.nil?
```

File: app/controllers/api/v1/projects_controller.rb [hunk 8]
```diff
@@ -0,0 +1,132 @@
+# frozen_string_literal: true
+
+class Api::V1::ProjectsController < Api::V1::BaseController
+  include ActionView::Helpers::SanitizeHelper
+
+  before_action :authenticate_user, only: [:index]
+  before_action :authenticate_user!, except: [:index]
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:7]: We need to show project view and index page without authorization

Comment (tachyons) [reply app/controllers/api/v1/projects_controller.rb:7]: Ie auth should be optional for project index and view pages

File: app/controllers/api/v1/projects_controller.rb [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    def set_project
+      @project = Project.find(params[:id])
+      @author = @project.author
+    end
+
+    def load_index_projects
+      @projects = if @current_user.nil?
+        Project.public_access
+      else
+        Project.where("author_id = ? OR  project_access_type = ?", @current_user.id, "Public")
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:99]: Use scope

File: app/controllers/api/v1/projects_controller.rb [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      end
+    end
+
+    def load_user_projects
+      @projects = if @current_user.id == params[:id].to_i
+        @current_user.projects
+      else
+        Project.where("author_id = ? AND project_access_type = ?", params[:id], "Public")
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:107]: Use scope

File: app/helpers/sorting_helper.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+module SortingHelper
+  def self.sort_fields(sort, allowed)
+    allowed = allowed.map(&:to_s)
+    fields = sort.to_s.split(",")
+    ordered_fields = ordered_hash(fields)
+    ordered_fields.select { |key, _value| allowed.include?(key) }
+  end
+
+  def self.ordered_hash(fields)
```
Comment (tachyons) [inline app/helpers/sorting_helper.rb:11]: Missing Spec for above two methods

**Comment by Nitish145** [comment]
@tachyons please check!!

File: app/controllers/api/v1/projects_controller.rb [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+  # GET /api/v1/projects/featured
+  def featured_circuits
+    @options[:links] = link_attrs(paginate(@projects), api_v1_projects_featured_url)
+    render json: Api::V1::ProjectSerializer.new(paginate(@projects), @options)
+  end
+
+  # GET /api/v1/projects/:id/toggle-star
+  def toggle_star
+    star = Star.find_by(user_id: @current_user.id, project_id: @project.id)
```
Comment (tachyons) [inline app/controllers/api/v1/projects_controller.rb:64]: Can we abstract this into project class
```ruby
class Project
  def toggle_star
  end
end
```

File: app/models/project.rb [hunk 13]
```diff
@@ -22,6 +23,10 @@ class Project < ApplicationRecord
   scope :public_and_not_forked,
         -> { where(project_access_type: "Public", forked_project_id: nil) }
 
+  scope :public_access, -> { where(project_access_type: "Public") }
```
Comment (tachyons) [inline app/models/project.rb:26]:
```suggestion
  scope :public, -> { where(project_access_type: "Public") }
```
What about just public? It reads better for me

File: app/models/project.rb [hunk 14]
```diff
@@ -22,6 +23,10 @@ class Project < ApplicationRecord
...
+
+  scope :author, ->(author_id) { where(author_id: author_id) }
```
Comment (tachyons) [inline app/models/project.rb:28]:
```suggestion
  scope :by, ->(author_id) { where(author_id: author_id) }
```

```ruby
 Project.author(author_id)
```

```ruby
 Project.by(author_id)
```
Second one seems express the context better

File: app/models/project.rb [hunk 15]
```diff
@@ -59,6 +64,15 @@ def increase_views(user)
     end
   end
 
+  def fork(user)
```
Comment (tachyons) [inline app/models/project.rb:67]: 👍

**Comment by tachyons** [comment]
Awesome 🎉

**Pull request merged**